### ISSUE TITLE
TypeVar for self breaks super

### ISSUE THREAD

**Issue by esoma** [python/mypy#11678]
```python
from typing import TypeVar

T = TypeVar('T')

class X:
    def __init__(self: T):
        super().__init__() # error: Argument 2 for "super" not an instance of argument 1
```

Possibly related to https://github.com/python/typing/discussions/909? But it seems the actual `super` call is handled specially so, possibly not?

It would seem to me if `self` is a `TypeVar` we can safely say that it must be compatible with super's arg 2...which seems like a simple fix. If `T` is bound to an incompatible type we'll get errors elsewhere to cover that situation (`error: The erased type of self "__main__.A" is not a supertype of its class "__main__.X"`).

**Comment by sobolevn** [comment]
@esoma yes, I also think that this should be supported. We can also check corner cases for type vars with bounds and explicit values.

Would you want to work on this? 🙂

**Comment by esoma** [comment]
@sobolevn Yeah, I'll look into it.

**Comment by esoma** [comment]
Found a test that works this by declaring:
```python
T = TypeVar('T', bound='X', covariant=True)
```

So maybe that's just the solution -- it works for my particular use case at least.

**Comment by ethanscorey** [comment]
Thanks for opening this issue — I've run into the same problem. Just wanted to chime in to say that I tried your solution of adding `covariant=True` to my `TypeVar` declaration, but it didn't work for me.

```python
from typing import TypeVar

from django.db import models


ModelType = TypeVar("ModelType", bound=models.Model, covariant=True)


class MyModel(models.Model):
    def save(self: ModelType, *args, **kwargs) -> None:
        ...
        super().save(*args, **kwargs)
```

**Comment by esoma** [comment]
@ethanscorey 
The bound target needs to be the subclass, not the base class:
```
ModelType = TypeVar("ModelType", bound='MyModel', covariant=True)
```
Unfortunately for every subclass you'll need a unique `TypeVar`.

**Comment by erictraut** [comment]
This appears to have been fixed. Mypy 1.5 doesn't generate any errors for the sample at the top of this issue.

**Comment by hauntsaninja** [comment]
Fixed in #13488
